routine/src/clj/parts/routine/core.clj

```diff
@@ -0,0 +1,136 @@
+  "Shutdowns a `java.util.concurrent.ExecutorService` gracefully. Based
+   on example `shutdownAndAwaitTermination` from
+   https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html"
+  [w]
+  ;; Disable new tasks from being submitted:
+  (let [^java.util.concurrent.ExecutorService scheduler (:scheduler w)]
+    (.shutdown scheduler)
+    (try
+      ;; Wait a while for existing tasks to terminate:
+      (when-not (.awaitTermination scheduler
+                                   termination-timeout
+                                   java.util.concurrent.TimeUnit/MILLISECONDS)
+        ;; Cancel currently executing tasks:
+        (.shutdownNow scheduler)
+        ;; Wait a while for tasks to respond to being cancelled:
+        (when-not (.awaitTermination scheduler
+                                     termination-timeout
+                                     java.util.concurrent.TimeUnit/MILLISECONDS)
+          ((:log/error w
+             identity)
+           {:log/error :scheduler/termination-failed
+            :scheduler scheduler})))
+      (catch InterruptedException e
+        ;; (Re-)Cancel if current thread also interrupted:
+        (.shutdownNow scheduler)
+        ;; In comparison to the example, do not interrupt the Thread
+        ;; that executes the system shutdown:
+        ;;
+        ;; (.interrupt (Thread/currentThread))
+        ))))
+
+(defn schedule-light-routines!
+  [w]
+  (let [light-routines (filter
+                         (fn [routine]
+                           (= (:routine/category routine) :light))
+                         (:routines w))
+        scheduled-routines (doall (map (fn [r]
+                                         (assoc r
+                                                :routine/scheduled-future
+                                                (schedule-routine! (assoc w
+                                                                          :scheduler (:routine/light-routine-scheduler w)
+                                                                          :routine r))))
+                                       light-routines))]
+    (assoc w
+           :routine/scheduled-light-routines
+           scheduled-routines)))
+
+(defn schedule-heavy-routines!
+  [w]
+  (let [heavy-routines (filter
+                         (fn [routine]
+                           (= (:routine/category routine) :heavy))
+                         (:routines w))
+        scheduled-routines (doall (map (fn [r]
+                                         (assoc r
+                                                :routine/scheduled-future
+                                                (schedule-routine! (:routine/heavy-routine-scheduler w)
+                                                                   r)))
+                                       heavy-routines))]
+    (assoc w
+           :routine/scheduled-heavy-routines
+           scheduled-routines)))
```

Comment:
`:routine/scheduled-heavy-routines` and `:routine/scheduled-light-routines` probably do not need to end up in the system map?